### PR TITLE
Concatenate all coords, labels and masks

### LINKED ISSUE #909: concatenate of dataset without data returns empty dataset
Consider a dataset `d` with only coords or labels. `concatenate(d,d, dim)` returns and *empty* dataset. I would have expected concatenation of the coords and labels. Is this expectation correct?

### PR DESCRIPTION
**Pull request by DanNixon**
Fixes #909

### CODE REVIEW

File: core/include/scipp/core/dimensions.h [hunk 1]
```diff
@@ -36,6 +37,10 @@ class SCIPP_CORE_EXPORT Dimensions {
     for (const auto [label, size] : dims)
       addInner(label, size);
   }
+  Dimensions(const std::unordered_map<Dim, scipp::index> &dims) {
+    for (const auto [label, size] : dims)
+      addInner(label, size);
+  }
```
Comment (SimonHeybrock) [inline core/include/scipp/core/dimensions.h]: I do not understand this. `Dimensions` implies a particular ordering of dimensions, where `std::unordered_map` is the opposite.

File: core/dataset_operations.cpp [hunk 2]
```diff
@@ -106,7 +106,13 @@ DataArray concatenate(const DataConstProxy &a, const DataConstProxy &b,
...
 Dataset concatenate(const DatasetConstProxy &a, const DatasetConstProxy &b,
                     const Dim dim) {
-  Dataset result;
+  const auto da = Dimensions(a.dimensions());
+  const auto db = Dimensions(b.dimensions());
+  Dataset result(std::map<std::string, Variable>(),
+                 concat(a.coords(), b.coords(), dim, da, db),
+                 concat(a.labels(), b.labels(), dim, da, db),
+                 concat(a.masks(), b.masks(), dim, da, db),
+                 std::map<std::string, Variable>());
```
Comment (SimonHeybrock) [inline core/dataset_operations.cpp]: I think I see why you added the constructor, but it has to be solved differently. Looking at `concat` it appears that it is not making any use of the features that require ordering. Can you template the `dimsA` and `dimsB` arguments in `concat` so it also accepts other types (such as `unordered_map`) that have `operator[]`?

File: core/test/concatenate_test.cpp [hunk 3]
```diff
@@ -211,3 +211,22 @@ TEST(ConcatenateTest, concatenate_sparse_no_data) {
...
+  b.setLabels("abc",
+              makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{.3, .4}));
+
+  const auto res = concatenate(a, b, Dim::X);
+
+  EXPECT_EQ(res.coords()[Dim::X],
+            makeVariable<double>(Dims{Dim::X}, Shape{4}, Values{1, 2, 3, 4}));
+  EXPECT_EQ(res.labels()["abc"], makeVariable<double>(Dims{Dim::X}, Shape{4},
+                                                      Values{.1, .2, .3, .4}));
+}
```
Comment (SimonHeybrock) [inline core/test/concatenate_test.cpp]: Can you add a test with an additional coord or label that represents bin edges?

File: core/include/scipp/core/dimensions.h (hunk 1 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/dimensions.h]: (see below)

File: core/test/concatenate_test.cpp (hunk 3 above)
Comment (DanNixon) [reply core/test/concatenate_test.cpp]: Without an associated data item how can you determine if a coordinate represents bin edges or not?

Comment (SimonHeybrock) [reply core/test/concatenate_test.cpp]: It looks at the dims of the other coords or labels.

**Pull request merged**